packages/web3-providers-ws/test/integration/web_socket_provider_integration.test.ts

```diff
@@ -0,0 +1,338 @@
+					webSocketProvider.on(
+						'message',
+						(
+							error: Error | null,
+							result?: JsonRpcSubscriptionResult | JsonRpcNotification<any>,
+						) => {
+							if (error) {
+								done.fail(error);
+							}
+							expect(result?.id).toBe(jsonRpcPayload.id);
+							done();
+						},
+					);
+					webSocketProvider.request(jsonRpcPayload).catch(err => {
+						done.fail(err);
+					});
+				});
+
+				it('should subscribe on error', done => {
+					const errorMsg = 'Custom WebSocket error occured';
+					webSocketProvider.on('error', (err: any) => {
+						expect(err?.message).toBe(errorMsg);
+						done();
+					});
+
+					webSocketProvider['_wsEventEmitter'].emit(
+						'error',
+						new Web3WSProviderError(errorMsg),
+					);
+				});
+
+				// eslint-disable-next-line jest/expect-expect
+				it('should subscribe on connect', done => {
+					webSocketProvider.on('open', () => {
+						done();
+					});
+				});
+
+				it('should subscribe on close', done => {
+					const code = 1001;
+					// const provider = new WebSocketProvider(clientWsUrl);
+					webSocketProvider.on(
+						'close',
+						(err: Error | null, event: OnCloseEvent | null | undefined) => {
+							if (err) {
+								done.fail(err);
+							}
+							expect(event!.code).toEqual(code);
+							done();
+						},
+					);
+					currentAttempt = 0;
+					waitForOpenConnection(webSocketProvider)
+						.then(() => {
+							webSocketProvider.disconnect(code);
+						})
+						.catch(() => {
+							done.fail();
+						});
+				});
+			});
+			describe('disconnect and reset test', () => {
+				it('should disconnect', async () => {
+					const provider = new WebSocketProvider(
+						clientWsUrl,
+						{},
+						{ delay: 1, autoReconnect: false, maxAttempts: 1 },
+					);
+					await waitForOpenConnection(provider);
+					provider.disconnect(1000);
+					await waitForOpenConnection(provider, 'disconnected');
+					expect(provider.getStatus()).toBe('disconnected');
+				});
+				it('should reset', () => {
+					jsonRpcPayload = {
+						jsonrpc: '2.0',
+						id: 42,
+						method: 'eth_getBalance',
+						params: [accounts[0], 'latest'],
+					} as Web3APIPayload<EthExecutionAPI, 'eth_getBalance'>;
+					const defPromise = new DeferredPromise<JsonRpcResponse<ResponseType>>();
+
+					const reqItem: WSRequestItem<any, any, any> = {
+						payload: jsonRpcPayload,
+						deferredPromise: defPromise,
+					};
+
+					webSocketProvider['_pendingRequestsQueue'].set(
+						jsonRpcPayload.id as JsonRpcId,
+						reqItem,
+					);
+					expect(webSocketProvider['_pendingRequestsQueue'].size).toBe(1);
+
+					webSocketProvider['_sentRequestsQueue'].set(
+						jsonRpcPayload.id as JsonRpcId,
+						reqItem,
+					);
+					expect(webSocketProvider['_sentRequestsQueue'].size).toBe(1);
+
+					webSocketProvider.reset();
+					expect(webSocketProvider['_pendingRequestsQueue'].size).toBe(0);
+					expect(webSocketProvider['_sentRequestsQueue'].size).toBe(0);
```

Comment:
In general one should not interact with private data fields or behaviors in the integration tests. If you think something is really necessary to test from user perspective and it does not have a public interface then you should create one.

Reply:
These queues are not needed to be exposed to the end-user, but, I want to be sure that they are emptied when `reset` is called. Any ideas on how this could be done without exposing them?

Reply:
The thumb rule is, If something is not exposed as public it should be tested in integration tests, instead in unit tests. Integration is tests simply test the user expectations and if something not exposed, means user is not aware of internal implementation. There could be a queue, or there could be a stack or something else as a user I would not be testing these details. 

Just in case you wonder a way, then one way is to convert those `private` to `protected` then while testing we create a dummy class inherited from there and create public methods which use the protected properties and expose them. And then you use that dummy class in tests. In Java and C# there is a concept of Reflections, which can be used to test such scenarios.